src/metabase/lib/join.cljc

```diff
@@ -580,11 +592,67 @@
      (when-let [fk-col (fk-column-for query stage-number pk-col)]
        (lib.filter/filter-clause (equals-join-condition-operator-definition) fk-col pk-col)))))
 
-(mu/defn joined-thing :- [:maybe [:or lib.metadata/TableMetadata lib.metadata/CardMetadata]]
+(def ^:private Joinable
+  [:or lib.metadata/TableMetadata lib.metadata/CardMetadata])
+
+(mu/defn joined-thing :- [:maybe Joinable]
   "Return metadata about the origin of `a-join` using `metadata-providerable` as the source of information."
   [metadata-providerable :- lib.metadata/MetadataProviderable
    a-join                :- ::lib.schema.join/join]
   (let [origin (-> a-join :stages first)]
     (cond
       (:source-card origin)  (lib.metadata/card metadata-providerable (:source-card origin))
       (:source-table origin) (lib.metadata/table metadata-providerable (:source-table origin)))))
+
+(defn- add-join-alias-to-joinable-columns [cols a-join]
+  (let [join-alias     (current-join-alias a-join)
+        unique-name-fn (lib.util/unique-name-generator)]
+    (mapv (fn [col]
+            (as-> col col
+              (with-join-alias col join-alias)
+              (add-source-and-desired-aliases a-join unique-name-fn col)))
+          cols)))
+
+(defn- mark-selected-joinable-columns
+  "Mark the column metadatas in `cols` as `:selected` if they appear in `a-join`'s `:fields`."
+  [cols a-join]
+  (let [j-fields (join-fields a-join)]
+    (case j-fields
+      :all        (mapv (fn [col]
+                          (assoc col :selected? true))
+                        cols)
+      (:none nil) (mapv (fn [col]
+                          (assoc col :selected? false))
+                        cols)
+      ;; figure out which columns are in `:fields`, and then match them to the closest match out of `all-source-refs`.
+      (let [selected-fields-refs (mapv lib.ref/ref j-fields)
+            ;; pre-calculate refs for all the cols so we can match them up to the ones in `:fields`.
+            cols                 (mapv (fn [col]
+                                         (assoc col ::ref (lib.ref/ref col)))
+                                       cols)
+            all-source-refs      (mapv ::ref cols)
+            selected-source-refs (into #{}
+                                       (map #(lib.equality/find-closest-matching-ref % all-source-refs))
+                                       selected-fields-refs)]
+        (mapv (fn [col]
+                (->  col
+                     (assoc :selected? (contains? selected-source-refs (::ref col)))
+                     (dissoc ::ref)))
+              cols)))))
```

Comment:
I think this is a recurring pattern, it might make sense to extract it into a function.

Reply:
Agree, but I'd rather do it in a follow on and get this in sooner